src/modules/sdexec/sdexec.c

```diff
@@ -0,0 +1,622 @@
+        flux_log (h, LOG_ERR, "error parsing [systemd] config: %s", error.text);
+        return -1;
+    }
+    if (!unit_prefix)
+        unit_prefix = "flux-";
+    if (!(cpy = strdup (unit_prefix))) {
+        flux_log (h, LOG_ERR, "out of memory");
+        return -1;
+    }
+    *unit_prefixp = cpy;
+    return 0;
+}
+
+static void sdexec_ctx_destroy (struct sdexec_ctx *ctx)
+{
+    if (ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        free (ctx->unit_prefix);
+        flux_future_destroy (ctx->f);
+        flux_msglist_destroy (ctx->kills);
+        flux_msglist_destroy (ctx->requests);
+        free (ctx);
+        errno = saved_errno;
+    }
+}
+
+static struct sdexec_ctx *sdexec_ctx_create (flux_t *h)
+{
+    struct sdexec_ctx *ctx;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+    if (flux_get_rank (h, &ctx->rank) < 0)
+        goto error;
+    if (!(ctx->requests = flux_msglist_create ())
+        || !(ctx->kills = flux_msglist_create ()))
+        goto error;
+    return ctx;
+error:
+    sdexec_ctx_destroy (ctx);
+    return NULL;
+}
+
+static void delete_message (struct flux_msglist *msglist,
+                            const flux_msg_t *msg)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        if (msg == m) {
+            flux_msglist_delete (msglist);
+            return;
+        }
+        m = flux_msglist_next (msglist);
+    }
+}
+
+static const flux_msg_t *lookup_message_bypath (struct flux_msglist *msglist,
+                                                const char *path)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        struct sdproc *proc = flux_msg_aux_get (m, "sdproc");
+        if (proc && streq (proc->path, path))
+            return m;
+        m = flux_msglist_next (msglist);
+    }
+    return NULL;
+}
+
+static const flux_msg_t *lookup_message_bypid (struct flux_msglist *msglist,
+                                               pid_t pid)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        struct sdproc *proc = flux_msg_aux_get (m, "sdproc");
+        if (proc && proc->pid == pid)
+            return m;
+        m = flux_msglist_next (msglist);
+    }
+    return NULL;
+}
+
+static const flux_msg_t *lookup_message_byaux (struct flux_msglist *msglist,
+                                               const char *name,
+                                               void *value)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        if (flux_msg_aux_get (m, name) == value)
+            return m;
+        m = flux_msglist_next (msglist);
+    }
+    return NULL;
+}
+
+static void start_properties_changed (struct sdexec_ctx *ctx,
+                                      const flux_msg_t *msg,
+                                      json_t *properties)
+{
```

Comment:
idea, document expected state transitions that can/should occur within this function and log unexpected state transitions?  B/c its systemd, some bad admin may do something that wouldn't be expected normally.

Reply:
Yes good idea!